### PR TITLE
Add controller-bootstrap go template files

### PR DESCRIPTION
**Pull request by JamesJHPark**
Issue: [aws-controllers-k8s/community/issues/1358](https://github.com/aws-controllers-k8s/community/issues/1358)

Description of changes:

- This PR is to add the various go template files which will be used to bootstrap a controller.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: template/.gitignore.tpl
Comment (RedbackThomson) [inline template/.gitignore.tpl]: This line in the ignore isn't necessary for this repo

File: template/GOVERNANCE.md.tpl
Comment (RedbackThomson) [inline template/GOVERNANCE.md.tpl]: Probably this entire section can be removed. Possibly we want to rewrite it to remain relevant but otherwise I don't think anyone will refer to this.

File: template/INSTRUCTIONS.md.tpl
Comment (RedbackThomson) [inline template/INSTRUCTIONS.md.tpl]: I might consider renaming this `READ_BEFORE_COMMIT.md` or something a little more explicit about it being used before submitting the first PR.

File: template/INSTRUCTIONS.md.tpl
Comment (RedbackThomson) [inline template/INSTRUCTIONS.md.tpl]: I don't understand what this step is asking the user to do. `Ignore the resource` sounds like it's saying don't take any action? And shouldn't the CLI have already run `make build-controller`?

File: template/INSTRUCTIONS.md.tpl
Comment (RedbackThomson) [inline template/INSTRUCTIONS.md.tpl]: e2e tests aren't quite as simple as add the tests inside the test directory. Probably better to refer them back to the testing documentation for this, rather than repeating ourselves here

File: template/INSTRUCTIONS.md.tpl
Comment (RedbackThomson) [inline template/INSTRUCTIONS.md.tpl]:
```suggestion
3. Add your team members' GitHub aliases to your team controller alias in [OWNERS_ALIASES](OWNERS_ALIASES) file.
```

File: template/Makefile.tpl
Comment (RedbackThomson) [inline template/Makefile.tpl:2]: Empty file on purpose?

File: template/apis/v1alpha1/README.md.tpl
Comment (RedbackThomson) [inline template/apis/v1alpha1/README.md.tpl]: This file seems pointless haha. I'm assuming this is a placeholder for the directory so that it's created and the code-generator works properly?

I think I'd rather we just fix the code-generator problem before we merge this PR. I've assigned [this issue](https://github.com/aws-controllers-k8s/community/issues/1361) to me

File: template/generator.yaml.tpl
Comment (RedbackThomson) [inline template/generator.yaml.tpl]: I don't think this line is achieving anything here?

File: template/metadata.yaml.tpl
Comment (RedbackThomson) [inline template/metadata.yaml.tpl]: I would prefer if we don't generate these values - or at least we should include instructions for the service teams to check and update them. The purpose of this file was that the developers had final say over the displayed names of their controllers. ServiceID is not always the service abbreviation, and is not always capitalised correctly.

File: template/metadata.yaml.tpl
Comment (RedbackThomson) [inline template/metadata.yaml.tpl]: We can generate these with `api_version: v1alpha1` and `status: available`

File: template/olm/olmconfig.yaml.tpl
Comment (RedbackThomson) [inline template/olm/olmconfig.yaml.tpl]: We use a standard email and name for all controllers - please copy the values from an existing `olmconfig.yaml`

File: template/templates/README.md.tpl
Comment (RedbackThomson) [inline template/templates/README.md.tpl]: If you are using `README.md` files to generate directories, but not for displaying anything useful, I would instead recommend using an empty file with the name `.gitkeep`. The purpose of `.gitkeep` is so that a directory is kept when committing to git, even if it has no contents.

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
This PR looks like it hits most of the files that we need. I can understand most of the substitutions, and have left some comments inline about places where we can be a bit smarter about those (or where we can't assume something). Almost everything is just in the details.

I see you've made an attempt at generating the `test/e2e` directory, but there are a lot of things missing from that which can be auto-generated. I am happy to talk you through that if you want to message me on Slack.

File: template/INSTRUCTIONS.md.tpl
Comment (vijtrip2) [inline template/INSTRUCTIONS.md.tpl]:
```suggestion
1. To get started, look inside the `generator.yaml` file, and remove the resource(s) from the ignore list and execute `make build-controller` from the ACK code-generator.
```

File: template/INSTRUCTIONS.md.tpl
Comment (vijtrip2) [inline template/INSTRUCTIONS.md.tpl]:
```suggestion
3. Add your team members to your team controller alias in `OWNERS_ALIASES` file.
```

File: template/Makefile.tpl
Comment (vijtrip2) [reply template/Makefile.tpl:2]: +1
Makefile should be generated for initial bootstrap.

File: template/OWNERS.tpl
Comment (vijtrip2) [inline template/OWNERS.tpl:4]:
```suggestion
approvers:
  - core-ack-team
  - service-team
```

File: template/OWNERS_ALIASES.tpl
Comment (vijtrip2) [inline template/OWNERS_ALIASES.tpl]:
```suggestion
  # TODO: Add your team members GitHub alias to the team alias
  service-team: []
```

File: template/templates/README.md.tpl
Comment (vijtrip2) [inline template/templates/README.md.tpl]:
```suggestion
templates directory is used to provide custom hook code templates.
```

Comment (vijtrip2) [reply template/templates/README.md.tpl]: Current information is not sufficient, but i believe we should keep README.md files to provide developer(without much ACK knowledge) some context about these directories.

File: template/test/README.md.tpl
Comment (vijtrip2) [inline template/test/README.md.tpl]:
```suggestion
test directory contains the python e2e tests for this controller.
```

**Review by vijtrip2 — COMMENTED** [review]
Great start, @JamesJHPark !

File: template/INSTRUCTIONS.md.tpl
Comment (JamesJHPark) [reply template/INSTRUCTIONS.md.tpl]: Thank you for the suggestion, this has been modified accordingly.

File: template/INSTRUCTIONS.md.tpl
Comment (JamesJHPark) [reply template/INSTRUCTIONS.md.tpl]: Thank you for pointing this out, I have revised the instruction to refer to the testing documentation.

File: template/INSTRUCTIONS.md.tpl
Comment (JamesJHPark) [reply template/INSTRUCTIONS.md.tpl]: This has been modified to: 1. To get started, look inside the `generator.yaml` file, and remove the resource(s) from the ignore list and execute `make build-controller` from the ACK code-generator.

File: template/Makefile.tpl
Comment (JamesJHPark) [reply template/Makefile.tpl:2]: Makefile has been added. Thank you.

File: template/test/README.md.tpl
Comment (JamesJHPark) [reply template/test/README.md.tpl]: The suggested changes have been implemented. Thank you!

**Review by a-hilaly — APPROVED** [review]
Neat! Good job @JamesJHPark !

File: template/READ_BEFORE_COMMIT.md.tpl
Comment (vijtrip2) [inline template/READ_BEFORE_COMMIT.md.tpl]: There should be a link to this testing documentation here.
@RedbackThomson do you know if we have documentation on how to write/structure python e2e tests?
I found this [one](https://aws-controllers-k8s.github.io/community/docs/contributor-docs/testing/) which talk about how to execute e2e tests.

File: template/metadata.yaml.tpl
Comment (vijtrip2) [inline template/metadata.yaml.tpl:3]: I think it can be useful to fill in this value for service team by default. and provide them instruction to update it here and also in READ_BEFORE_COMMIT.md
```suggestion
  full_name: "{{ .ServiceFullName }}" [Update Accordingly]
  short_name: "{{ .ServiceID }}" [Update Accordingly]
```

File: template/READ_BEFORE_COMMIT.md.tpl
Comment (vijtrip2) [inline template/READ_BEFORE_COMMIT.md.tpl]: include instruction to update full_name and short_name inside metadata.yaml

Comment (JamesJHPark) [reply template/READ_BEFORE_COMMIT.md.tpl]: The instruction to update full_name and short_name inside metadata.yaml has been added. Thank you.

File: cmd/controller-bootstrap/command/root.go
Comment (RedbackThomson) [inline cmd/controller-bootstrap/command/root.go]: Please rename to `test-infra-commit-sha`, as we don't want to prescribe that it be the HEAD  of the tree (and accordingly change the description).

File: Makefile
Comment (RedbackThomson) [inline Makefile]: Given how many flags there are now, I might suggest that we use the long form for everything (`--model-name`, `--service-name`, etc.). I think short codes are great when you're doing stuff quickly on the command line, but in scripts we want the full transparency around what is what

File: template/test/e2e/tests/__init__.py
Comment (RedbackThomson) [inline template/test/e2e/tests/__init__.py]: Is it fine that this is a `.py` file? Should it be `.py.tpl`?

File: Makefile
Comment (JamesJHPark) [reply Makefile]: Thank you for the suggestion, the flags have been modified to use long form.

File: template/templates/README.md.tpl
Comment (RedbackThomson) [reply template/templates/README.md.tpl]: That's fine, but these README's aren't doing that. Plus this is another thing we now need to keep consistent across 50+ repositories. I'd rather we have contributor documentation on the website, where we can manage it all from one place, rather than replicating it everywhere

File: template/READ_BEFORE_COMMIT.md.tpl
Comment (RedbackThomson) [reply template/READ_BEFORE_COMMIT.md.tpl]: Nothing on the public internet. I have notes that I've kept since working with Bruce on the RDS team which would be good to publish

File: template/metadata.yaml.tpl
Comment (RedbackThomson) [reply template/metadata.yaml.tpl:3]: I'm very scared to do that because I don't trust that people will have a high bar for this sort of detail. This is something I hope we will catch in the code review, but the casing is pretty important here. These strings are used in the documentation and some of the more user-facing parts of the system, so we need to keep them consistent and "pretty".

File: template/READ_BEFORE_COMMIT.md.tpl
Comment (RedbackThomson) [inline template/READ_BEFORE_COMMIT.md.tpl]: I think these instructions need to be a lot more detailed. We need to tell them not only what, but why and how. This will more easily let us consistently share information for new repository owners and reduce the amount we have to repeat ourselves.

```suggestion
1. To get started, edit `generator.yaml` file and comment out (with a `#`) each of the resource(s) from the ignore list, generating your controller after every edit.

  This step will enable each resource within the service to be generated into an ACK custom resource definition. As you generate each of them, the code-generator may require additional configuration (in the `generator.yaml` file) in order for you to continue.
  
2. Update fields under `service` inside `metadata.yaml`

  This file provides display information about the controller and the service it supports. This file is intended to be manually updated to ensure it matches the proper casing and terminology for the service. For `full_name`, provide the full display name of the service (eg. `Amazon Simple Storage Service`). For `short_name`, provide the abbreviation or shortened service name (eg. `S3` or `SageMaker` - do not include `AWS` or `Amazon`). For `link`, provide a link to the homepage for the service. For `documentation`, provide a link to the homepage for the service documentation/user guide.
  
3. Add your team members' GitHub aliases to the `OWNERS_ALIASES` file

  This file determines who has permissions to review and merge code within the repository.

4. Add any custom code to the `templates` directory

  As you work on the controller, you may need to add custom hooks to support functionality specific to your service. Hook code for any resource should be placed in a file with the path `templates/hooks/<resource name>/<hook name>.go.tpl`
  
5. Refer to the open source documentation on testing for creating and running tests, which should be added in the `test/e2e` directory.

  All PRs containing changes to CRDs should be accompanied with appropriate tests for any new functionality or custom code. Documentation for these tests can be found in the "Contributor Docs" section of the ACK website.
```

Comment (JamesJHPark) [reply template/READ_BEFORE_COMMIT.md.tpl]: Yes, thank you for the insight and pointing this out. The instructions have been updated with the suggested details and information.

File: cmd/controller-bootstrap/command/root.go
Comment (RedbackThomson) [inline cmd/controller-bootstrap/command/root.go]: Does this need to be a flag, even? We can check that the files already exists and run the different routine conditionally? The code shouldn't do anything destructive, anyway, so even if someone runs this on an existing repository it won't ruin anything.

File: cmd/controller-bootstrap/command/root.go
Comment (RedbackThomson) [inline cmd/controller-bootstrap/command/root.go]: nit: Call this `output-path` or `output-directory`?

**Comment by vijtrip2** [comment]
Excellent work, @JamesJHPark !!
And super feedback @RedbackThomson !!
/lgtm

**Pull request merged**
